### PR TITLE
テスト: gitUtils と githubUtils のエッジケースのユニットテストカバレッジの向上

### PR DESCRIPTION
**Pull request by is0692vs**
This PR adds extensive edge-case unit test coverage for `src/gitUtils.ts` and `src/githubUtils.ts` to increase the overall project Codecov metrics.

## Changes:
- **`src/githubUtils.ts`**: Abstracted the dynamic `import('@octokit/rest')` into an exported `getOctokitInstance` helper. This simplifies the testing strategy by allowing Sinon stubs without relying on `proxyquire` which fails with ESM/CJS interop issues in this environment.
- **`src/test/githubUtils.unit.test.ts`**: Added missing coverage for URL parsing logic, branch info extraction (including handling deleted forks and successful resolution), and mocked branch creation calls.
- **`src/test/gitUtils.unit.test.ts`**: Added test logic to cover the failure state cases of missing the VS Code Git extension, missing workspace folders, missing remotes, and `undefined` fallback repositories.
- Re-ran the local test suite using `xvfb-run -a pnpm test` and `npx nyc pnpm run test:unit` to ensure no tests were broken and the expected coverage increased.

---
*PR created automatically by Jules for task [16728366965229602733](https://jules.google.com/task/16728366965229602733) started by @is0692vs*

<details open><summary><h3>Greptile Summary</h3></summary>

このPRは `githubUtils.ts` に DI 可能な `setOctokitFactory` ヘルパーと `OctokitClient` インターフェースを追加し、`gitUtils` / `githubUtils` のユニットテストカバレッジを拡充します。プロダクションコードの変更は妥当ですが、`githubUtils.unit.test.ts` の `createRemoteBranch` スイート内で **存在しない `setOctokitFactoryForTesting`（正: `setOctokitFactory`）が 3 箇所呼び出されており、該当テストが実行時に `TypeError` で失敗します**。
</details>

<details open><summary><h3>Confidence Score: 4/5</h3></summary>

プロダクションコードは問題ないが、テストの 3 ケースが実行時エラーで失敗するため、修正後にマージを推奨。

P1 の実行時 TypeError が `createRemoteBranch` の 3 テストに存在するため上限は 4。プロダクションコード側は正しく実装されており、他のテストも正常に動作する見込み。

src/test/githubUtils.unit.test.ts の 197・215・233 行目
</details>

<details open><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| src/githubUtils.ts | `OctokitClient` インターフェース・`setOctokitFactory` DI ヘルパーを追加し、`createRemoteBranch` / `getPullRequestBranchInfo` の Octokit 初期化をファクトリ経由に統一。ESM 互換で型安全な設計。 |
| src/test/githubUtils.unit.test.ts | `createRemoteBranch` スイートの 3 テスト（197・215・233行目）で存在しない `setOctokitFactoryForTesting` を呼び出しており、実行時に `TypeError` が発生してテストが失敗する。 |
| src/test/gitUtils.unit.test.ts | テストをネストされたスイート構造に整理し、`getGitApi` / `getRepositoryForWorkspaceFolder` / `getRemoteUrl` / `getCurrentBranchSha` の主要パスをカバー。`mockOutputChannel` のフル実装も追加済み。 |

</details>

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant Test as Unit Test
    participant GHUtils as githubUtils.ts
    participant Factory as octokitFactory (module-level)
    participant Octokit as @octokit/rest

    Note over Test,Octokit: プロダクション時
    GHUtils->>Factory: octokitFactory(token)
    Factory->>Octokit: import('@octokit/rest') + new Octokit()
    Octokit-->>GHUtils: OctokitClient

    Note over Test,Octokit: テスト時 (setOctokitFactory でDI)
    Test->>GHUtils: setOctokitFactory(mockFactory)
    GHUtils->>Factory: octokitFactory = mockFactory
    Test->>GHUtils: createRemoteBranch / getPullRequestBranchInfo
    GHUtils->>Factory: octokitFactory(token)
    Factory-->>GHUtils: mockOctokitClient
    GHUtils-->>Test: result
```

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
This is a comment left during a code review.
Path: src/test/githubUtils.unit.test.ts
Line: 197

Comment:
**存在しない関数 `setOctokitFactoryForTesting` の呼び出しによるテスト失敗**

`githubUtils.setOctokitFactoryForTesting(...)` を呼び出していますが、`src/githubUtils.ts` がエクスポートしているのは `setOctokitFactory` であり、`setOctokitFactoryForTesting` は存在しません。このため、このテストを実行すると `TypeError: githubUtils.setOctokitFactoryForTesting is not a function` が発生してテストが失敗します。同じ問題が 215 行目・233 行目にも存在します。

```suggestion
            githubUtils.setOctokitFactory(async () => mockOctokit as any);
```

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: src/test/githubUtils.unit.test.ts
Line: 215

Comment:
**同上: `setOctokitFactoryForTesting` は未エクスポート**

197 行目と同様に `setOctokitFactoryForTesting` を呼び出していますが、正しくは `setOctokitFactory` です。

```suggestion
            githubUtils.setOctokitFactory(async () => mockOctokit as any);
```

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: src/test/githubUtils.unit.test.ts
Line: 233

Comment:
**同上: `setOctokitFactoryForTesting` は未エクスポート**

197 行目・215 行目と同様に `setOctokitFactoryForTesting` を呼び出していますが、正しくは `setOctokitFactory` です。

```suggestion
            githubUtils.setOctokitFactory(async () => mockOctokit as any);
```

How can I resolve this? If you propose a fix, please make it concise.
`````

</details>

<sub>Reviews (2): Last reviewed commit: ["test: fix octokit factory test hook usag..."](https://github.com/hiroki-org/jules-extension/commit/4ede612c67a7a5066962a56c4463f18effab5378) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=29952424)</sub>

> Greptile also left **1 inline comment** on this PR.

### CODE REVIEW

File: src/githubUtils.ts [hunk 1]
```diff
@@ -23,14 +23,20 @@ export function parseGitHubUrl(url: string): GitHubUrlInfo | null {
...
+
 export async function createRemoteBranch(
     pat: string,
     owner: string,
     repo: string,
     branchName: string
 ): Promise<void> {
-    const { Octokit } = await import('@octokit/rest');
-    const octokit = new Octokit({ auth: pat });
+    const octokit = await exports.getOctokitInstance(pat);
```
[the remark this answers is not in the document]
Comment (is0692vs) [reply src/githubUtils.ts]: 対応しました: `exports.getOctokitInstance(...)` 呼び出しを廃止し、モジュール内の Octokit ファクトリ（DI）経由で取得する実装に変更しました。影響範囲: `createRemoteBranch` の Octokit 初期化経路。検証: `pnpm run check-types`, `pnpm run lint`, `pnpm run test:unit`.

File: src/githubUtils.ts [hunk 2]
```diff
@@ -86,8 +92,7 @@ export async function getPullRequestBranchInfo(
     prNumber: number
 ): Promise<PullRequestBranchInfo | null> {
     try {
-        const { Octokit } = await import('@octokit/rest');
-        const octokit = new Octokit({ auth: token });
+        const octokit = await exports.getOctokitInstance(token);
```
[the remark this answers is not in the document]
Comment (is0692vs) [reply src/githubUtils.ts]: 対応しました: `getPullRequestBranchInfo` でも `exports` 参照を廃止し、同じ DI ファクトリ経由で Octokit を取得するよう統一しました。影響範囲: PR ブランチ情報取得ロジック。検証: `pnpm run check-types`, `pnpm run lint`, `pnpm run test:unit`.

File: src/test/gitUtils.unit.test.ts [hunk 3]
```diff
@@ -1,216 +1,130 @@
...
+        });
+
+        test('should return repository if found', () => {
+            const repoMock = { rootUri: { fsPath: '/test/path' } };
+            const git = { repositories: [repoMock] };
+            const workspaceFolder = { uri: { fsPath: '/test/path' } } as any;
+            const repo = getRepositoryForWorkspaceFolder(git, workspaceFolder, mockOutputChannel);
+            assert.strictEqual(repo, repoMock);
+        });
+    });
```
[the remark this answers is not in the document]
Comment (is0692vs) [reply src/test/gitUtils.unit.test.ts:64]: 対応しました: 削除されていたログインジェクション回帰テストを `gitUtils.unit.test.ts` に復元し、改行を含むパス入力時にログ文字列へ実改行が混入しないことを検証するケースを追加しました。影響範囲: `getRepositoryForWorkspaceFolder` のログサニタイズ検証。検証: `pnpm run check-types`, `pnpm run lint`, `pnpm run test:unit`.

File: src/test/githubUtils.unit.test.ts [hunk 4]
```diff
@@ -1,65 +1,158 @@
...
+        test('should parse ssh url correctly', () => {
+            const url = 'git@github.com:owner/repo.git';
+            const result = githubUtils.parseGitHubUrl(url);
+            assert.deepStrictEqual(result, { owner: 'owner', repo: 'repo' });
+        });
+        test('should return null for invalid urls', () => {
+            assert.strictEqual(githubUtils.parseGitHubUrl('invalid-url'), null);
+            assert.strictEqual(githubUtils.parseGitHubUrl('https://gitlab.com/owner/repo'), null);
+        });
     });
```
[the remark this answers is not in the document]
Comment (is0692vs) [reply src/test/githubUtils.unit.test.ts:36]: 対応しました: `parseGitHubUrl` のエッジケーステストを再追加しました（`http://`、`.git` なし SSH、ハイフン/ドットを含む owner/repo、`/blob/...` で null）。影響範囲: `src/test/githubUtils.unit.test.ts`。検証: `pnpm run check-types`, `pnpm run lint`, `pnpm run test:unit`.

File: src/githubUtils.ts [hunk 5]
```diff
@@ -23,14 +23,20 @@ export function parseGitHubUrl(url: string): GitHubUrlInfo | null {
         repo: match[2],
     };
 }
+// Helper factory function to abstract dynamic Octokit import and instantiation.
+// This makes it easy to stub in unit tests without complex module mocking or proxyquire.
+export async function getOctokitInstance(token: string): Promise<any> {
+    const { Octokit } = await import('@octokit/rest');
+    return new Octokit({ auth: token });
+}
```
[the remark this answers is not in the document]
Comment (is0692vs) [reply src/githubUtils.ts]: 対応しました: `getOctokitInstance` の戻り値を `any` から、実際に利用する API 面を持つ `OctokitClient` 型に変更しました。あわせてテスト用にファクトリ差し替え関数を追加し、型安全性を保ったままモックできるようにしています。影響範囲: `src/githubUtils.ts` と関連ユニットテスト。検証: `pnpm run check-types`, `pnpm run lint`, `pnpm run test:unit`.

File: src/githubUtils.ts [hunk 6]
```diff
@@ -23,14 +23,20 @@ export function parseGitHubUrl(url: string): GitHubUrlInfo | null {
         repo: match[2],
     };
 }
+// Helper factory function to abstract dynamic Octokit import and instantiation.
+// This makes it easy to stub in unit tests without complex module mocking or proxyquire.
+export async function getOctokitInstance(token: string): Promise<any> {
```
[the remark this answers is not in the document]
Comment (is0692vs) [reply src/githubUtils.ts]: 対応しました: `getOctokitInstance` を `Promise<any>` から具象 API 面を持つ型（`OctokitClient`）で定義し、呼び出し側での型情報を維持するようにしました。影響範囲: `src/githubUtils.ts`。検証: `pnpm run check-types`, `pnpm run lint`, `pnpm run test:unit`.

File: test_octokit_dynamic.js [hunk 7]
```diff
@@ -0,0 +1,2 @@
+const { getOctokitInstance } = require('./out/githubUtils.js');
+getOctokitInstance('dummy').then(o => console.log(typeof o)).catch(console.error);
```
[the remark this answers is not in the document]
Comment (is0692vs) [reply test_octokit_dynamic.js]: 対応しました: `out/` 依存の手動確認スクリプト（`test_octokit_dynamic.js` と map）を削除し、確認観点は `src/test/githubUtils.unit.test.ts` のユニットテストに寄せました。影響範囲: テスト資産整理のみ。検証: `pnpm run check-types`, `pnpm run lint`, `pnpm run test:unit`.

File: src/githubUtils.ts [hunk 8]
```diff
@@ -23,14 +23,63 @@ export function parseGitHubUrl(url: string): GitHubUrlInfo | null {
...
+    octokitFactory = factory;
+}
+
+export function resetOctokitFactoryForTesting(): void {
+    octokitFactory = defaultOctokitFactory;
+}
+
+export async function getOctokitInstance(token: string): Promise<OctokitClient> {
+    return octokitFactory(token);
+}
```
[the remark this answers is not in the document]
Comment (is0692vs) [reply src/githubUtils.ts]: ご指摘ありがとうございます。コメント対象は outdated であることを確認しました。

本PRでは、既存方針どおり setup/teardown で factory を毎回 reset してリークを防止する形を維持します。非公開化やDI再設計は別PRで扱います。

File: src/test/githubUtils.unit.test.ts [hunk 9]
```diff
@@ -1,65 +1,204 @@
...
+                sha: '1234567890abcdef'
+            });
+        });
+
+        test('should throw error if getOctokitInstance throws', async () => {
+            githubUtils.setOctokitFactoryForTesting(async () => {
+                throw new Error('Auth failed');
+            });
+            await assert.rejects(githubUtils.createRemoteBranch('token', 'owner', 'repo', 'new-branch'), /Auth failed/);
+        });
```
[the remark this answers is not in the document]
Comment (is0692vs) [reply src/test/githubUtils.unit.test.ts:185]: 対応しました。createRemoteBranch の downstream API 失敗系を追加しています。
- repos.get が reject
- git.getRef が reject
- git.createRef が reject

反映コミット: dc48891

**Comment by is0692vs** [comment]
@greptile

File: src/test/githubUtils.unit.test.ts [hunk 10]
```diff
@@ -1,65 +1,241 @@
...
+            const mockOctokit = {
+                repos: {
+                    get: sandbox.stub().rejects(new Error('Failed to get repository'))
+                },
+                git: {
+                    getRef: sandbox.stub().resolves({ data: { object: { sha: '1234567890abcdef' } } }),
+                    createRef: sandbox.stub().resolves()
+                }
+            };
+            githubUtils.setOctokitFactory(async () => mockOctokit as any);
```
[the remark this answers is not in the document]
Comment (is0692vs) [reply src/test/githubUtils.unit.test.ts:197]: 対応済みです。setOctokitFactoryForTesting をすべて setOctokitFactory に置き換えました。反映コミット: 4ede612

[the remark this answers is not in the document]
Comment (is0692vs) [reply src/test/githubUtils.unit.test.ts:197]: 今回は対応見送りとします。理由: 現在の実装では `src/githubUtils.ts` 側が `setOctokitFactory` を export しており、`src/test/githubUtils.unit.test.ts` も同名 API を呼び出す状態に揃っているため、指摘の `setOctokitFactoryForTesting is not a function` は再現しません（ローカルで `pnpm run check-types`, `pnpm run lint`, `pnpm run test:unit` も通過）。代替策/前提: もし CI 上で同エラーが再発する場合は、該当 run の checkout SHA とログをもとに再確認します。
